### PR TITLE
Optionally zoom towards the cursor

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
This PR introduces an option to zoom towards the cursor rather than the centre of the screen. If disabled, which it is by default, vanilla behaviour is retained.

### CODE REVIEW

**Comment by IntelOrca** [comment]
The code looks fine, but I think you should change the behaviour to instead zoom in with the point under the cursor remain still under the cursor. This is what OpenTTD does.

Basically when you hover over an object on the side of your screen and zoom in from way out to way in, your cursor should still be on that object.

**Comment by AaronVanGeffen** [comment]
You're right, that would make it much more intuitive. I'll see whether I can implement the behaviour you describe in a nice manner.

**Comment by AaronVanGeffen** [comment]
I've given it another go, but it's not quite right yet. This is mostly notable on the outer zoom levels, that is, less so zooming in.

Am I interpreting the viewport coordinate system correctly?

**Comment by janisozaur** [comment]
Fixes https://github.com/OpenRCT2/OpenRCT2/issues/1141

**Comment by AaronVanGeffen** [comment]
Figured it out. Push incoming after I've patched things up a little. :)

**Comment by AaronVanGeffen** [comment]
Figured it out. I wasn't increasing the offset with the zoom level, and still had a placeholder for the base height. Mending both results in smooth behaviour, I think. 😄

Currently, I have switched it off by default. What are your thoughts on this? I do want to make it optional, but I wonder whether it'll help people find the feature if it's on by default...

**Comment by IntelOrca** [comment]
@AaronVanGeffen We can make it on by default, this way everyone will be exposed to it. We'll know soon enough if people don't like it.

**Comment by AaronVanGeffen** [comment]
Done. Cheers for the quick feedback, by the way.

**Comment by mrtnptrs** [comment]
I would like to test this before you merge it to catch bugs, but appveyor did not check this PR strangely, so I have no access to a working build including these changes. (and no, I'm not able to build OpenRCT2 myself currently.) Why wasn't this PR checked by appveyor?

**Comment by AaronVanGeffen** [comment]
This hasn't been merged yet as it needs some more work in terms of accuracy. There is a slight offset due to not taking into account the position of the cursor *within* the tile currently pointing to, which becomes really apparent when zooming in and out a few times without moving the cursor.

AppVeyor did check earlier commits, including acd3ba4, but it hasn't done bf73f2e, no. The change in that commit was trivial, though, so it's not that big a deal.

**Comment by AaronVanGeffen** [comment]
@IntelOrca Per 8ac1b79, the offset of the cursor relative to the tile it is pointing to is now computed and applied when setting zoom level. The code can be optimised still, I'm sure, but I'd first like some input.

This commit resolves the inconsistencies mentioned on Gitter, *unless* the map itself is offset when zooming out. This most notably happens when zooming all the way out. I'm not sure how to tackle this, if at all. Any ideas?

**Comment by IntelOrca** [comment]
> the offset of the cursor relative to the tile it is pointing to is now computed and applied when setting zoom level. The code can be optimised still, I'm sure, but I'd first like some input.

I think its working fine now. I am happy to merge it now unless you want to do anything else first?

> This commit resolves the inconsistencies mentioned on Gitter, unless the map itself is offset when zooming out. This most notably happens when zooming all the way out. I'm not sure how to tackle this, if at all. Any ideas?

I think that's a different problem related to viewport boundaries, that can be fixed separately from this.

**Comment by AaronVanGeffen** [comment]
Cheers. I'm satisfied enough with it per 29b0b48, so please merge it, then.

File: src/interface/window.c [hunk 1]
```diff
@@ -1445,6 +1445,49 @@ void window_rotate_camera(rct_window *w, int direction)
...
+	// Compute map coordinate by mouse position.
+	get_map_coordinates_from_pos(mouse_x, mouse_y, VIEWPORT_INTERACTION_MASK_NONE, map_x, map_y, NULL, NULL, NULL);
+
+	// Get viewport coordinates centring around the tile.
+	int base_height = map_element_height(*map_x, *map_y);
+	int dest_x, dest_y;
+	center_2d_coordinates(*map_x, *map_y, base_height, &dest_x, &dest_y, w->viewport);
+
+	// Rebase mouse position onto centre of window, and compensate for zoom level.
+	int rebased_x = ((w->width >> 1) - mouse_x) << w->viewport->zoom,
```
Comment (janisozaur) [inline src/interface/window.c]: @AaronVanGeffen you cannot left-shift negative values, this is undefined behaviour. Can you fix this (and others in this function)?

Comment (AaronVanGeffen) [inline src/interface/window.c]: @janisozaur Right, I see what you mean. You'd rather I use simple multiplication and division here, then?

Comment (janisozaur) [inline src/interface/window.c]: Yes

Comment (AaronVanGeffen) [inline src/interface/window.c]: https://github.com/OpenRCT2/OpenRCT2/pull/4249 deals with this. Thanks.

**Pull request merged**